internal/unionstore/art/art_iterator.go

```diff
@@ -156,140 +156,161 @@ func (it *Iterator) init(lowerBound, upperBound []byte) {
-				if it.inner.compare(helper) < 0 || len(helper.idxes) == 0 {
-					it.valid = false
-				}
-				return
-			}
-			it.valid = false
-		}
-		return
+		startKey, endKey = upperBound, lowerBound
 	}
 
-	it.inner.idxes, it.inner.nodes = it.seek(lowerBound)
-	if len(upperBound) == 0 {
-		it.endAddr = arena.NullAddr
+	if len(startKey) == 0 {
+		it.inner.seekToFirst(it.tree.root, it.reverse)
 	} else {
-		helper := new(baseIter)
-		helper.allocator = &it.tree.allocator
-		helper.idxes, helper.nodes = it.seek(upperBound)
-		if it.inner.compare(helper) < 0 {
-			// upperBound is exclusive, so we move the helper cursor to the previous node, which is the true endAddr.
-			it.endAddr = helper.prev().addr
-			if it.inner.compare(helper) > 0 || len(helper.idxes) == 0 {
-				it.valid = false
-			}
-			return
-		}
+		it.inner.seek(it.tree.root, startKey)
+	}
+	if len(endKey) == 0 {
+		it.endAddr = arena.NullAddr
+		return
+	}
+
+	helper := new(baseIter)
+	helper.allocator = &it.tree.allocator
+	helper.seek(it.tree.root, endKey)
+	cmp := it.inner.compare(helper)
+	if cmp == 0 {
+		// no keys exist between start key and end key, set the iterator to invalid.
 		it.valid = false
 		return
 	}
+
+	if it.reverse {
+		it.endAddr = helper.next().addr
+	} else {
+		it.endAddr = helper.prev().addr
+	}
+	cmp = it.inner.compare(helper)
+	if cmp == 0 {
+		// the current key is valid.
+		return
+	}
+	// in asc scan, if cmp > 0, it means current key is larger than end key, set the iterator to invalid.
+	// in desc scan, if cmp < 0, it means current key is less than end key, set the iterator to invalid.
+	if cmp < 0 == it.reverse || len(helper.idxes) == 0 {
+		it.valid = false
+	}
+}
+
+// baseIter is the inner iterator for ART tree.
+// You need to call seek or seekToFirst to initialize the iterator.
+// after initialization, you can call next or prev to iterate the ART.
+// next or prev returns nullArtNode if there is no more leaf node.
+type baseIter struct {
+	allocator *artAllocator
+	// the baseIter iterate the ART tree in DFS order, the idxes and nodes are the current visiting stack.
+	idxes []int
+	nodes []artNode
+}
+
+// seekToFirst seeks the boundary of the tree, sequential or reverse.
+func (it *baseIter) seekToFirst(root artNode, reverse bool) {
+	// if the seek key is empty, it means -inf or +inf, return root node directly.
+	it.nodes = []artNode{root}
+	if reverse {
+		it.idxes = []int{node256cap}
+	} else {
+		it.idxes = []int{inplaceIndex}
+	}
 }
 
-// seek the first node and index that >= key, return the indexes and nodes of the lookup path
-// nodes[0] is the root node
-func (it *Iterator) seek(key artKey) ([]int, []artNode) {
-	curr := it.tree.root
+// seek the first node and index that >= key, nodes[0] is the root node
+func (it *baseIter) seek(root artNode, key artKey) {
+	curr := root
 	depth := uint32(0)
-	idxes := make([]int, 0, 8)
-	nodes := make([]artNode, 0, 8)
+	it.idxes = make([]int, 0, 8)
+	it.nodes = make([]artNode, 0, 8)
```

Comment:
Better to do the allocation lazily before its usages.

And move the check
```
if len(key) == 0 {
    panic(...)
}
```
above